Custom resource schema modifications

danielrbradley opened the pull request:
Add a more understandable and unified approach to modifying existing parts of the schema.

Use a function to define a resource's schema–resource & types in both the schema and metadata. This is given any existing declaration which it is allowed to modify.

- If newly added types conflict with an existing type in the schema, we will raise an error.
- Allows modifying any of the types associated with the custom resource.
- Allow modification of both the schema and the metadata in one abstration.
- Allows creation of brand new resources - the existing resource will just be nil.

This allows the removal of the custom resource `Types`, `Schema` and `Meta` fields along with the associated `SchemaMixins`, `SchemaTypeMixins` and `MetaMixins` functions.

danielrbradley commented:
> * I'm not sure why we need to discover all referenced types upfront, since most custom resources won't even use them. Could we offer `findTypesForResource` as a utility for custom resources instead, or even get away with just `TypeLookupFunc` for resources that know which types they want to access?

It felt like an easier API if you get passed values and you return values rather than callbacks. It avoids questions like "what happens if they call the callback lots of times" etc too. The types will only be searched if the custom resource implements the `SchemaModifications` function (which is only the Dashbaord right now). 

> * When two custom resources touch the same resource or type, the result depends on the order of iteration `customresources.ApplySchemaModifications`. It's sufficiently unlikely so it's ok, I think, but even better would be to keep track of modifications and warn if something is touched a second time.

Yes, good catch on the map iteration. Interestingly this is also the case for all the mixin functions too. I'll change them all to iterate by name for consistency at least.

> * Do we need `MetaTypes` when `SchemaModifications` can also add metadata?

Not strictly, but it's a simpler form to use if only adding and not modifying anything existing.

thomas11 commented:
> > * I'm not sure why we need to discover all referenced types upfront, since most custom resources won't even use them. Could we offer `findTypesForResource` as a utility for custom resources instead, or even get away with just `TypeLookupFunc` for resources that know which types they want to access?
> 
> It felt like an easier API if you get passed values and you return values rather than callbacks. It avoids questions like "what happens if they call the callback lots of times" etc too. The types will only be searched if the custom resource implements the `SchemaModifications` function (which is only the Dashbaord right now).

Makes sense!
 
> > * Do we need `MetaTypes` when `SchemaModifications` can also add metadata?
> 
> Not strictly, but it's a simpler form to use if only adding and not modifying anything existing.

In isolation, it's simpler, but having two ways to do the same thing is not simpler IMO (think of debugging).

danielrbradley commented:
Rebased and squashed reworking following review .. and implemented the dashboard changes.

mikhailshilkov commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -0,0 +1,93 @@
+				Description: "Dashboard part settings.",
+			}
+			resource.Types["azure-native:portal:DashboardPartMetadata"] = dashbordPartMetadata
+			// The response type is the same as the input type (there's no refs)
```
How do you know there are no refs in other markdown part properties? (in the future)

mikhailshilkov commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -0,0 +1,93 @@
+			// The response type is the same as the input type (there's no refs)
+			resource.Types["azure-native:portal:DashboardPartMetadataResponse"] = dashbordPartMetadata
+
+			// Mirror the changes in the metadata
```
Any possibility to reduce duplication and factor the changes into a reusable function?

mikhailshilkov commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -0,0 +1,93 @@
+				AdditionalProperties: &resources.AzureAPIProperty{Ref: "pulumi.json#/Any"},
+			}
+			resource.MetaTypes["azure-native:portal:DashboardPartMetadata"] = dashbordPartMetadataMeta
+			resource.MetaTypes["azure-native:portal:DashboardPartMetadataResponse"] = dashbordPartMetadataMeta
```
Should we try to avoid hard-coding all these magic strings and build them instead?

danielrbradley replied:
The existing string builder functions were really difficult to read in order to see what we're actually creating in the schema. I think just writing out the strings is pretty clear to see what it will do.

mikhailshilkov commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -0,0 +1,93 @@
+			// Add a new MetadataType to replace the MarkdownPartMetadata type
+			dashboardParts := resource.Types["azure-native:portal:DashboardParts"]
+			dashboardPartsMetadata := dashboardParts.Properties["metadata"]
+			dashboardPartsMetadata.Ref = "#/types/azure-native:portal:DashboardPartMetadata"
```
Should we assert on the current Ref value? In case the spec changes and doesn't point to MarkdownPart anymore?

mikhailshilkov reviewed (commented):
This looks like a step in the right direction, thank you! Should we check for possibilities to rewrite some existing schema modifications in this style?

mjeffryes reviewed (commented):
I agree with the general direction here, but there are a lot of details to sort out still. I suspect, it will be easier to reason about as a pure refactor, independent of restoring the dashboard customization. 

Would like to see this fully replace all the custom edits in `genMixins` as well.

danielrbradley commented on provider/pkg/resources/customresources/customresources.go:
```diff
@@ -50,6 +56,81 @@ type CustomResource struct {
+}
+
+// ApplySchemas applies custom schema modifications to the given package.
+// These modifications should never overlap with each other, but we apply in a deterministic order to ensure
```
Yup, what Matt said - we _shouldn't_ ever have two custom resources modifying the same types, but in case we ever did, let's make sure it doesn't change every time we generate. I think trying to add validation that types aren't modified by two resources is tricky to write and not worth the time given how few custom resources we have (which don't overlap).

thomas11 replied:
Making sure the output is deterministic will just hide the bug, then. If anything, shouldn't we do the opposite and randomize so the issue becomes visible?

danielrbradley replied:
I think the issue is how you identify what's causing the flip-flopping - it took us years to identify the similar issues in the examples generation.

mikhailshilkov replied:
How was this comment addressed?

danielrbradley commented on provider/pkg/resources/customresources/customresources.go:
```diff
@@ -50,6 +56,81 @@ type CustomResource struct {
+// ApplySchemas applies custom schema modifications to the given package.
+// These modifications should never overlap with each other, but we apply in a deterministic order to ensure
+// that the end result of the modifications is consistent.
+func ApplySchemas(pkg *pschema.PackageSpec, meta *resources.AzureAPIMetadata) error {
```
It's not only modifications. This function can also be used to create new resources and types. I think this will be clearer once the legacy functions are removed.

mikhailshilkov replied:
I'd love for names to make sense for their current purpose, not for a future idea. Once the "legacy" functions are removed, we can rename.

danielrbradley replied:
IMO this makes sense for what it can currently do, even if it's not yet being used for it. The only confusing part is that the legacy code exists at all. I've renamed `Schema` to `LegacySchema` and `SchemaF` to `Schema` for better forward naming.

danielrbradley commented on provider/pkg/resources/customresources/customresources.go:
```diff
@@ -50,6 +56,81 @@ type CustomResource struct {
+		existingResource, ok := pkg.Resources[r.tok]
+		var originalResource *ResourceDefinition
+		types := map[string]pschema.ComplexTypeSpec{} // Hoist scope for easy lookup when checking if the type is new.
+		if ok {
```
Exactly - the new schema function covers both modifying existing schemas as well as defining new schemas. Can rename the `ok` for clarity.

danielrbradley commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -0,0 +1,93 @@
+	return &CustomResource{
+		tok: "azure-native:portal:Dashboard",
+		SchemaF: func(resource *ResourceDefinition) (*ResourceDefinition, error) {
+			if resource == nil {
```
This is by design to allow this to be called for creating new resources too.

danielrbradley commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -0,0 +1,93 @@
+				// This should only happen if we're running with a namespace or version filter for testing, so we'll allow it to pass.
+				return nil, nil
+			}
+			// Add a new MetadataType to replace the MarkdownPartMetadata type
```
MetadataType is the data used by the provider at runtime which should align to the schema types.

danielrbradley commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -0,0 +1,93 @@
+			dashboardPartsMetadata.Ref = "#/types/azure-native:portal:DashboardPartMetadata"
+			dashboardParts.Properties["metadata"] = dashboardPartsMetadata
+			resource.Types["azure-native:portal:DashboardParts"] = dashboardParts
+			// Do the same for the response type
```
Sure, wasn't clear how to do this well initially as each operation is on a different type - not keen on some heavy generics.

danielrbradley commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -0,0 +1,93 @@
+// We override both input and output (Response) types.
+func portalDashboard() *CustomResource {
+	return &CustomResource{
+		tok: "azure-native:portal:Dashboard",
```
Not exactly sure how we iterate the set of available versions right now. I'd suggest we just leave this as-is for now – just fixing the default version.

danielrbradley commented on provider/pkg/resources/customresources/customresources.go:
```diff
@@ -50,6 +56,81 @@ type CustomResource struct {
+		if err != nil {
+			return fmt.Errorf("failed to apply custom resource schema modifications for %s: %w", r.tok, err)
+		}
+		if customResource == nil {
```
A nil currently means no modification will be made. This is only really used currently when debugging generation and we don't recieve a source resource when we're expecting one. The ideal here would be to log a warning, but we don't have any kind of diagnostics framework implemented to allow this easily.

I'll just document the current behaviour and we can always adapt it later as needed.

danielrbradley commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -0,0 +1,93 @@
+			dashboardPartsResponse.Properties["metadata"] = dashboardPartsMetadataResponse
+			resource.Types["azure-native:portal:DashboardPartsResponse"] = dashboardPartsResponse
+
+			// Find the old MarkdownPartMetadata and derive the new DashboardPartMetadata from it
```
I'll switch approach so we're more similar to the existing implementation then.

danielrbradley commented on provider/pkg/resources/customresources/customresources.go:
```diff
@@ -50,6 +56,81 @@ type CustomResource struct {
+// ApplySchemas applies custom schema modifications to the given package.
+// These modifications should never overlap with each other, but we apply in a deterministic order to ensure
+// that the end result of the modifications is consistent.
+func ApplySchemas(pkg *pschema.PackageSpec, meta *resources.AzureAPIMetadata) error {
```
The new implementation mirrors the test approach of the existing generation code - which is covered largely via snapshot testing.

mikhailshilkov replied:
Given snapshot tests haven't changed in this PR, I don't see how they cover this code.

danielrbradley replied:
I've pushed the snapshot test

danielrbradley commented:
I've rebased, addressed the minor comments and reworked the Portal schema to be much more similar to the existing code, with the addition of asserts over each type we're replacing to flag when there's changes upstream which affect the parts we're overwriting.

mikhailshilkov commented on provider/pkg/resources/customresources/customresources.go:
```diff
@@ -30,11 +33,14 @@ type CustomResource struct {
+	// The resource and types (both schema and metadata) returned will be written back into the schema and metadata.
+	// This can also include new types which were not present in the original schema.
+	// Any dangling, unreferenced types will be removed from the schema automatically.
+	SchemaF func(resource *ResourceDefinition) (*ResourceDefinition, error)
```
Are you planning to move all custom resources to this? Looks like a surprising interface for pure custom resources.

danielrbradley replied:
Yes, because it's then a single function to define all the parts of the schema rather than various separate properties which are amalgamated into a single resource.

mikhailshilkov reviewed (commented):
The changes since my last review look good. Still missing test coverage.

mikhailshilkov commented on provider/pkg/resources/customresources/customresources.go:
```diff
@@ -102,6 +191,7 @@ func BuildCustomResources(env *azureEnv.Environment,
 		portalDashboard(),
 		customWebApp,
 		customWebAppSlot,
+		portalDashboard(),
```
Just noticed you added it for the second time here

danielrbradley replied:
👍 Fixed

mikhailshilkov commented on provider/pkg/gen/gen_dashboard_test.go:
```diff
@@ -0,0 +1,37 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	snaps.MatchSnapshot(t, generationResult.Schema)
```
In my opinion, snapshot tests are no substitute to proper unit tests. It was okay to add them post-factum for all the existing code that we had no tests for, but I suggest we don't use them as an excuse not to write tests for net-new non-trivial functionality.

Your current PR is a nice illustration of my case: as far as I can tell, your snapshot is actually not testing anything at all. The only `portal` resource that the snapshot contains is `azure-native:portal/v20200901preview:Dashboard`, which is not processed by the custom resource you implemented, so it just passes `nil` to the `Schema` method, and the method exists immediately. Yet, the "test" provides an impression that we are testing something, and it's hard to tell otherwise without looking into the details.

@thomas11 @mjeffryes I'd love to get your opinion on this testing approach and whether it matches mine or not.

danielrbradley replied:
Yes, it looks like the snapshot currently misses generating the default resource, I'll take a look at fixing that when I get opportunity.

@mikhailshilkov Perhaps if you're envisioning something specific with the unit tests that you think is more effective, you could contribute it?

thomas11 replied:
Apart from the snaps vs unit tests question, a test that passes without asserting anything is a bug. Can we add some assertions here to prevent that? Along the lines of "`generationResult` contains foo".

In general, I'm all for unit tests over snap-based tests, but for things the portal resource where you have a bunch of opaque JSON, snaps can still a good fit.

mikhailshilkov replied:
How can we test anything except the happy path with snapshot tests? I guess we miss a mechanism to amend snapshots to cause errors/edge cases and assert the outcomes?

thomas11 replied:
My understanding is that snapshot tests, as we're using them, are strictly for the happy path.

mikhailshilkov commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -119,6 +168,39 @@ func dashboardPartMetadataType() schema.ComplexTypeSpec {
+					TypeSpec: schema.TypeSpec{
+						Type: "string",
+					},
+					Description: "The dashboard part metadata type.\nExpected value is 'Extension/HubsExtension/PartType/MarkdownPart'.",
```
On the second pass, I'm not sure I like the strictness of this. We will have to re-write this code in case upstream changes the description of a type that we are throwing away, right? That seems unnecessary fragile?

danielrbradley replied:
This is why the original design I proposed attempted to derrive the modified schema from the original schema in order to be more flexible, but the feedback on that was that it was too complex & fragile. The frequency of updates hasn't been too high previously so we can always refine these asserts if needed later.

mikhailshilkov commented on provider/pkg/resources/customresources/custom_portal_dashboard.go:
```diff
@@ -31,10 +32,32 @@ func portalDashboard() *CustomResource {
 
+			existingMarkdownPartMetadataType := resource.Types["azure-native:portal:MarkdownPartMetadata"]
+			if !reflect.DeepEqual(existingMarkdownPartMetadataType, expectedMarkdownPartMetadataType("")) {
+				return nil, fmt.Errorf("unexpected azure-native:portal:MarkdownPartMetadata type: %#v", existingMarkdownPartMetadataType)
```
Should we provide guidance to future maintainers about what they need to do in this situation?

danielrbradley replied:
Sure, can refine these message further, though having any messages is already a big improvement over the current failure mode.

thomas11 commented on provider/pkg/gen/schema_test.go:
```diff
@@ -33,17 +33,29 @@ func TestTypeAliasFormatting(t *testing.T) {
 var _ Versioning = (*versioningStub)(nil)
 
 type versioningStub struct {
+	shouldIncline   func(provider string, version string, typeName, token string) bool
```
"Incline"?

danielrbradley replied:
Oops, yes - should be include. Ended up taking out the code which calls this mock. Will rename as I rebase 👍

mikhailshilkov commented:
Note that the "Dashboard Fix" section of the description was not correct anymore, so I deleted it
